lib/vnet/vnet.go

```diff
@@ -367,13 +442,134 @@ func (m *Manager) assignTCPHandler(handler tcpHandler) (tcpip.Address, error) {
 	addr := ipv6WithSuffix(m.ipv6Prefix, u32ToBytes(ipSuffix))
 
 	m.state.tcpHandlers[addr] = handler
+	m.state.appIPs[fqdn] = addr
 	if err := m.addProtocolAddress(addr); err != nil {
 		return addr, trace.Wrap(err)
 	}
 
 	return addr, nil
 }
 
+func (m *Manager) handleUDP(req *udp.ForwarderRequest) {
+	m.wg.Add(1)
+	go func() {
+		defer m.wg.Done()
+		m.handleUDPConcurrent(req)
+	}()
+}
+
+func (m *Manager) handleUDPConcurrent(req *udp.ForwarderRequest) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	id := req.ID()
+	slog := m.slog.With("request", id)
+	slog.DebugContext(ctx, "Handling UDP request.")
+	defer slog.DebugContext(ctx, "Finished handling UDP request.")
+
+	handler, ok := m.getUDPHandler(id.LocalAddress)
+	if !ok {
+		slog.DebugContext(ctx, "No handler for address.")
+		return
+	}
+
+	var wq waiter.Queue
+	endpoint, err := req.CreateEndpoint(&wq)
+	if err != nil {
+		slog.ErrorContext(ctx, "Failed to create UDP endpoint.", "error", err)
+		return
+	}
+
+	conn := gonet.NewUDPConn(m.stack, &wq, endpoint)
+	defer conn.Close()
+
+	if err := handler.HandleUDP(ctx, conn); err != nil {
+		slog.DebugContext(ctx, "Error handling UDP conn.", "error", err)
+	}
+}
+
+func (m *Manager) getUDPHandler(addr tcpip.Address) (UDPHandler, bool) {
+	m.state.mu.RLock()
+	defer m.state.mu.RUnlock()
+	handler, ok := m.state.udpHandlers[addr]
+	return handler, ok
+}
+
+func (m *Manager) assignUDPHandler(addr tcpip.Address, handler UDPHandler) error {
+	m.state.mu.Lock()
+	defer m.state.mu.Unlock()
+	if _, ok := m.state.udpHandlers[addr]; ok {
+		return trace.AlreadyExists("Handler for %s is already set", addr)
+	}
+	if err := m.addProtocolAddress(addr); err != nil {
+		return trace.Wrap(err)
+	}
+	m.state.udpHandlers[addr] = handler
+	return nil
+}
+
+// ResolveA implements [dns.Resolver.ResolveA].
+func (m *Manager) ResolveA(ctx context.Context, fqdn string) (dns.Result, error) {
+	result, err := m.resolveAAAA(ctx, fqdn)
+	if err != nil {
+		return dns.Result{}, trace.Wrap(err)
+	}
+	if result.AAAA != ([16]byte{}) {
+		// Matched a known app but not supporting IPv4 yet, return NoRecord.
+		return dns.Result{
+			NoRecord: true,
+		}, nil
+	}
+	return result, nil
+}
+
+// ResolveAAAA implements [dns.Resolver.ResolveAAAA].
+func (m *Manager) ResolveAAAA(ctx context.Context, fqdn string) (dns.Result, error) {
+	result, err := m.resolveAAAA(ctx, fqdn)
+	return result, trace.Wrap(err)
+}
+
+func (m *Manager) resolveAAAA(ctx context.Context, fqdn string) (dns.Result, error) {
```

Comment:
Just a few narrative comments below which I used to aid in my own understanding when reviewing, take or leave.

Reply:
those look helpful, i'll take them, thanks